core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -270,27 +323,90 @@ class LogSegment(val log: FileMessageSet,
     catch {
       case e: IOException => throw kafkaStorageException("index", e)
     }
+    try timeIndex.renameTo(new File(CoreUtils.replaceSuffix(timeIndex.file.getPath, oldSuffix, newSuffix)))
+    catch {
+      case e: IOException => throw kafkaStorageException("timeindex", e)
+    }
+  }
+
+  /**
+   * Append the last time index entry to the time index when this log segment become inactive segment.
```

Comment:
To be more precise, we are not adding the last, but the largest time index entry.